### PR TITLE
Feat/es7

### PR DESCRIPTION
**Pull request by thanh-nguyen-dang**
Jira Ticket: [PXP-10748](https://ctds-planx.atlassian.net/browse/PXP-10748)

### Breaking Changes

- Any environment using ElasticSearch 6.x will break with the new version

### Deployment changes
- We need to upgrade the target environment to use ElasticSearch 7.10 or OpenSearch before using this new guppy.
- We also need to upgrade Tube to new version that is compatible with ES7. Checking [this link](https://github.com/uc-cdis/tube/pull/238) for the corresponding PR of Tube.

[PXP-10748]: https://ctds-planx.atlassian.net/browse/PXP-10748?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ocshawn** [comment]
when attempting to run:
`GUPPY_PORT=3000 INTERNAL_LOCAL_TEST=true npm start`
i get the following error: 
`[09:22:25] INFO: [ES.initialize] getting array fields from es config index "gen3-dev-config"...
[09:22:25] ERROR: [ES.initialize] wrong array entry from config index: index "gen3-dev-subject" not found, skipped.
[09:22:25] ERROR: [ES.initialize] wrong array entry from config index: index "gen3-dev-file" not found, skipped.
[09:22:25] INFO: [ES.initialize] got array fields from es config index: {}
/var/git/guppy/src/server/schema.js:73
  const result = Object.keys(properties).map((field) => {
                        ^

TypeError: Cannot convert undefined or null to object`

**Comment by thanh-nguyen-dang** [comment]
> when attempting to run: `GUPPY_PORT=3000 INTERNAL_LOCAL_TEST=true npm start` i get the following error:
> 
> ```
> [09:22:25] INFO: [ES.initialize] getting array fields from es config index "gen3-dev-config"...
> [09:22:25] ERROR: [ES.initialize] wrong array entry from config index: index "gen3-dev-subject" not found, skipped.
> [09:22:25] ERROR: [ES.initialize] wrong array entry from config index: index "gen3-dev-file" not found, skipped.
> [09:22:25] INFO: [ES.initialize] got array fields from es config index: {}
> /var/git/guppy/src/server/schema.js:73
>   const result = Object.keys(properties).map((field) => {
>                         ^
> 
> TypeError: Cannot convert undefined or null to object
> ```
> 
> this is after re running setup as defined in https://github.com/uc-cdis/guppy/blob/master/devHelper/README.md#step3-start-server-for-developing-server-side-code

You need to use ES7. I think you local ES is @ocshawn

**Comment by ocshawn** [comment]
i ran this command `docker-compose -f ./devHelper/docker/esearch.yml up -d`

it looks like that file sets up an old version, please update

**Comment by thanh-nguyen-dang** [comment]
> i ran this command `docker-compose -f ./devHelper/docker/esearch.yml up -d`
> 
> it looks like that file sets up an old version, please update

I fixed the docker compose issue @ocshawn . Thanks for letting me know the issue

**Review by ocshawn — DISMISSED** [review]
looks good to me, i only tested locally; setup and storybook look and work great

please fill out PR details if this is a breaking change and any deployment details that are needed

File: src/server/es/aggs.js [hunk 1]
```diff
@@ -581,7 +581,7 @@ export const textAggregation = async (
   // don't add missing alias to numeric field by default
   // since the value of missing alias is a string
   if (config.esConfig.aggregationIncludeMissingData && !isNumericField) {
-    missingAlias = { missing: config.esConfig.missingDataAlias };
```
Comment (mfshao) [inline src/server/es/aggs.js:584]: why is this changing? this `missingDataAlias` should be configurable via config

**Comment by mfshao** [comment]
there is currently a ES6 dockerfile in the dev script in master now so you should remove it for this PR

File: src/server/es/aggs.js (hunk 1 above)
Comment (mfshao) [reply src/server/es/aggs.js:584]: if guppy no longer returns `no data` (or the missing data alias as defined in confog) for missing data, some envs may be affected, for example https://github.com/uc-cdis/gitops-qa/blob/master/qa-covid19.planx-pla.net/portal/gitops.json#L149 this one depends on that missing data alias value

We need to make sure this is tested

**Comment by mfshao** [comment]
when I was trying to run `./generate_data.sh` i got `TypeError: Cannot read properties of undefined (reading 'rawType')` 🤔

**Comment by thanh-nguyen-dang** [comment]
> when I was trying to run `./generate_data.sh` i got `TypeError: Cannot read properties of undefined (reading 'rawType')` 🤔

I did not change functions in genData.js @mfshao

**Comment by mfshao** [comment]
> > > > when I was trying to run `./generate_data.sh` i got `TypeError: Cannot read properties of undefined (reading 'rawType')` 🤔
> > > 
> > > 
> > > I did not change functions in genData.js @mfshao
> > 
> > 
> > right, so my question is if this related to ES7 changes, and if so, some changes might be needed for that function as well
> 
> There will be some changes needed. It has not been there because we don't use that when we implement the change. As I understand, it is just a util function and not affect to the correct of the running service

Yes but it is a part of the dev helper suite, which is also part of the functionalities of guppy, so it can help with dev work
Since it is also part of Guppy, we should update it with this ES7 PR as well

**Comment by thanh-nguyen-dang** [comment]
> > > > > when I was trying to run `./generate_data.sh` i got `TypeError: Cannot read properties of undefined (reading 'rawType')` 🤔
> > > > 
> > > > 
> > > > I did not change functions in genData.js @mfshao
> > > 
> > > 
> > > right, so my question is if this related to ES7 changes, and if so, some changes might be needed for that function as well
> > 
> > 
> > There will be some changes needed. It has not been there because we don't use that when we implement the change. As I understand, it is just a util function and not affect to the correct of the running service
> 
> Yes but it is a part of the dev helper suite, which is also part of the functionalities of guppy, so it can help with dev work Since it is also part of Guppy, we should update it with this ES7 PR as well

It has been resolved @mfshao

File: devHelper/docker/esearch.yml [hunk 2]
```diff
@@ -3,9 +3,7 @@ version: "3.3"
 services:
   # see https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#docker-cli-run-prod-mode
   elasticsearch:
-    image: elasticsearch
-    build:
-      dockerfile: es6.Dockerfile
```
Comment (mfshao) [inline devHelper/docker/esearch.yml:8]: this file should be removed form this directory

File: src/server/__mocks__/mockESData/mockNestedAggs.js [hunk 3]
```diff
@@ -58,7 +59,7 @@ const mockNestedAggs = () => {
             },
             {
               key: {
-                'visits.visit_label': 'no data',
```
Comment (mfshao) [inline src/server/__mocks__/mockESData/mockNestedAggs.js:61]: why is this behavior changing? this should retured the value of `missing_data_alias` from the guppy config file, if not configured then the default value should be `'no data'`

File: src/server/es/aggs.js (hunk 1 above)
Comment (mfshao) [reply src/server/es/aggs.js:584]: I think I have commented here but got no answer, can you elabrate? @thanh-nguyen-dang

Comment (thanh-nguyen-dang) [reply src/server/es/aggs.js:584]: ES does not support missing configuration anymore. We need to use different way {missing_bucket: true,  order: 'desc'} to configure the field with missing data. 
This part will handle the missing data
https://github.com/uc-cdis/guppy/pull/200/files#diff-5db64ea5481e180cbf163669ddb74578c387c32bafcbb7a4dccc8d57c1dc1102R685-R689

File: src/server/__mocks__/mockESData/mockNestedAggs.js (hunk 3 above)
Comment (thanh-nguyen-dang) [reply src/server/__mocks__/mockESData/mockNestedAggs.js:61]: ES does not support missing configuration anymore. We need to use different way {missing_bucket: true,  order: 'desc'} to configure the field with missing data. 
This part will handle the missing data
https://github.com/uc-cdis/guppy/pull/200/files#diff-5db64ea5481e180cbf163669ddb74578c387c32bafcbb7a4dccc8d57c1dc1102R685-R689

**Review by mfshao — APPROVED** [review]
looks good, thanks 🎉 
when making a release, please make sure to update the major version number since it is a breaking update (ES6 no longer being supported)

**Pull request merged**